### DISCUSSION TITLE
Use 'import' with ESM packages instead

### DISCUSSION THREAD

**Discussion by LNFWebsite** [vercel/next.js#35217] [Help] [answered]
I'm trying to import a library which switched to ESM and am getting this error:

```
Module not found: ESM packages (remark) need to be imported. Use 'import' to reference the package instead. https://nextjs.org/docs/messages/import-esm-externals
  2 | import path from 'path'
  3 | //import matter from 'gray-matter'
> 4 | import remark from 'remark'
    |                           ^
  5 | import html from 'remark-html'
```
Even though the module is being imported correctly.

When I change it to a name which doesn't exist I get:

```
./lib/posts.js:4:28
Module not found: Can't resolve 'remarke'
  2 | import path from 'path'
  3 | //import matter from 'gray-matter'
> 4 | import remark from 'remarke'
    |                            ^
  5 | import html from 'remark-html'
```

So I know that it is imported.

I also tried adding this to next.config.js:

```
experimental: {
        esmExternals: "loose",
    },
```

But then I get this error:

```
TypeError: (0 , _remark).default is not a function
```

This has something to do with the switch to ESM module of the remark library.

Does anyone know what I can do to get rid of this error?

**Comment by icyJoseph** [comment]
I am not sure if this would work, but have you tried, https://nextjs.org/blog/next-11-1#es-modules-support

```js
// next.config.js
module.exports = {
  // Prefer loading of ES Modules over CommonJS
  experimental: { esmExternals: true }
}
```

**Reply by LNFWebsite** [reply]
Just tried that and I get the same "Use 'import' to reference the package instead." error

**Reply by icyJoseph** [reply]
Ah, I'd try with this package https://www.npmjs.com/package/next-transpile-modules, I think I have a remark project somewhere, so I could test it later. Otherwise would be nice if someone figured this out

**Reply by LNFWebsite** [reply]
Thanks for the help. I might just end up downgrading remark until this gets fixed

**Reply by LNFWebsite** [reply]
This worked for anyone else who finds this discussion:

```
npm uninstall remark remark-html
npm install remark@13 remark-html@13
```

I found someone else was having a similar issue here: https://github.com/vercel/next.js/discussions/27758

**Reply by icyJoseph** [reply]
Well using latest Next.js I do not get this error...  using remark@14.0.2 it works out of the box.

```js
import {remark} from 'remark';
```

What version of Next.js do you use? As far as I understand this issue was fixed last year.

**Reply by LNFWebsite** [reply]
Next.js v12.1.0 and remark 14.0.2

I just deleted my package-lock.json, deleted node_modules, deleted .next and re-ran an npm install and I'm still getting the error.

The same error appears on trying to import micromark. I'm running node v14.15.3 Gonna try upgrading that rn and see if anything changes.

**Reply by LNFWebsite** [reply]
Just updated to node v16.14.0 and cleared everything again but same error

**Reply by LNFWebsite** [reply]
Ok I've made some progress. This is a use-case specific issue.

I'm importing remark in a file (/lib/posts.js) and then importing that file to a page (/pages/index.js).

This is when I'm getting the error. Otherwise, a straight import of remark in the page works fine.

**Reply by LNFWebsite** [reply]
Bingo! Fixed it.

The class file was exporting via

```
module.exports = Posts
```

which I guess somehow conflicts with loading ESM modules down the road. I switched it to

```
export default class Posts {
...
}
```

And it's working now.

**Reply by icyJoseph** [reply]
Awesome! Yeah I see, perhaps you needed to do `export = Posts` also, iirc, but I'd say it is best to have the same `import/export` syntax across your own modules.

**Reply by LNFWebsite** [reply]
Thanks for the help!

**Reply by icyJoseph** [reply]
If you could please mark this as a suggested answer, or write a small summary of what you did as a new entry and mark that as an answer :) You never know when someone else might have the same issue.

**Comment by LNFWebsite** [comment, marked as answer]
This fixed it: https://github.com/vercel/next.js/discussions/35217#discussioncomment-2343861

**Comment by TudorEsan** [comment]
I still had problems with the marked answer but this solved it:

https://nextjs.org/docs/messages/import-esm-externals
in next.config 
```
  experimental: {
    esmExternals: 'loose',
  },
```
